<p><img src="https://files.cdn.thinkific.com/file_uploads/133292/images/a9e/c1a/bc3/P02.jpg"></p>

`MCSDSC02P01V3______`

`MCSDSC02P01V2______`

`MCSDSELIMC02P01V1010621`

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Отлично! Проект принят с первого раза!
Оставил несколько комментариев ниже.

## Постановка задачи

Необходимо проанализировать данные сервиса предоставленного менеджером проекта. Данные предоставленые для опреденной территории и включает в себя некоторые данные по работе объектов находящихся на этой территории. 

*Пославленные задачи для анализа:*
>* На какие кафе приходится больше всего заказов? А на какие дни?
>* Чаще ли приложением пользуются именно в обеденное время?
>* Насколько большими компаниями  (или наоборот поодиночке) приходят в кафе?

## Импорт данных и библиотек

In [191]:
import pandas as pd

Прочитаем файл `bez_ocheredi.csv` и сохраним его в переменную `bo_data`.

In [219]:
bo_data = pd.read_csv("../../data/bez_ocheredi.csv")

Получим первые 10 строк таблицы.

In [220]:
bo_data.head(10)

userID    ресторан     weekday  ordered    Столик для 
0  b912c3c6a6ad   Iskra Bar     Monday     10:00             1
1  2dac0150717d   Iskra Bar     Monday     13:30             2
2  ba83aefab5dc   Iskra Bar    Tuesday     12:30             3
3  02996f1dd2ea   Iskra Bar  Wednesday     12:00             2
4  7c90be56d2ab   Iskra Bar   Thursday     14:30             3
5  51e0962387f7      Там Ем     Friday     12:30             2
6  c4f260a2b48a         NaN   Saturday     15:00             6
7  6b71309d6a8a  Крафтейнер     Monday     10:30             1
8  c5b969cb63a2  Бар Сибирь  Wednesday     15:30             2
9  b1570962e68c  Бар Сибирь  Wednesday     12:30             3

Получим общую информацию о данных из таблицы.

In [196]:
bo_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1024 entries, 0 to 1023
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   userID        1024 non-null   object
 1    ресторан     959 non-null    object
 2    weekday      1024 non-null   object
 3    ordered      1024 non-null   object
 4    Столик для   1024 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 40.1+ KB


Столбцы в `bo_data` содержат следующую информацию:

* userId - id пользователя 
* ресторан - название ресторана в котором был сделан заказ
* weekday - день недели в котором был сделан заказ
* ordered - время создание заказа
* столик для - количество мест за столиком на который был сделан заказ

### Выводы

* Таблица с данными состоит из 5 (4 не нулевых столбца типа object и 1 ненулевой столбец типа int64) столбцов и 1024 строк
* Некоторые строки содержат пропуски типа NaN
* Название колонок не однородные: имеют пробелы, некоторые колонки написаны на русском языке.

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Всё верно!

## Предобработка данных

Получим названия столбцов.

In [222]:
bo_data.columns

Index(['userID ', ' ресторан ', ' weekday', ' ordered ', ' Столик для '], dtype='object')

Обращаем внимание, что помимо названий на разном языке, в них есть пробелы, которые затрудняют доступ к данным. Переименуем столбцы и проверим результат.

In [223]:
bo_data.set_axis(['user_id','restaurant','weekday','ordered_time','seats_number'],axis='columns',inplace=True)
bo_data

user_id  restaurant    weekday ordered_time  seats_number
0     b912c3c6a6ad   Iskra Bar     Monday        10:00             1
1     2dac0150717d   Iskra Bar     Monday        13:30             2
2     ba83aefab5dc   Iskra Bar    Tuesday        12:30             3
3     02996f1dd2ea   Iskra Bar  Wednesday        12:00             2
4     7c90be56d2ab   Iskra Bar   Thursday        14:30             3
...            ...         ...        ...          ...           ...
1019  4b136de2c16a   Iskra Bar   Saturday        11:30             2
1020  9e46b91cc547  Бар Сибирь   Thursday        10:00             2
1021  5ca79042c44d      Там Ем   Thursday        14:00             3
1022  e0f9dfd1726c      Там Ем    Tuesday        17:30             1
1023  401d5fa8e55d      Там Ем  Wednesday        20:30             1

[1024 rows x 5 columns]

Еще раз проверим данные на наличие пропусков, просуммировав их.

In [224]:
bo_data.isnull().sum()

user_id          0
restaurant      65
weekday          0
ordered_time     0
seats_number     0
dtype: int64

Менеджер проекта предупреждал, что некоторые заказы заводились на тестовую карточку кафе без названия, поэтому есть смысл заполнить такой пропуск словом admin.

In [225]:
bo_data.fillna('admin', inplace=True)

Проверяем, что пропусков не осталось.

In [226]:
bo_data.isna().sum()

user_id         0
restaurant      0
weekday         0
ordered_time    0
seats_number    0
dtype: int64

Установим количество явных дубликатов. Если найдутся, то удалим и снова проверим, что их не осталось.

In [234]:
bo_data.duplicated().sum()

24

In [235]:
bo_data.drop_duplicates(inplace=True)
bo_data.reset_index(drop=True)

user_id  restaurant    weekday ordered_time  seats_number
0    b912c3c6a6ad   Iskra Bar     Monday        10:00             1
1    2dac0150717d   Iskra Bar     Monday        13:30             2
2    ba83aefab5dc   Iskra Bar    Tuesday        12:30             3
3    02996f1dd2ea   Iskra Bar  Wednesday        12:00             2
4    7c90be56d2ab   Iskra Bar   Thursday        14:30             3
..            ...         ...        ...          ...           ...
995  4b136de2c16a   Iskra Bar   Saturday        11:30             2
996  9e46b91cc547  Бар Сибирь   Thursday        10:00             2
997  5ca79042c44d      Там Ем   Thursday        14:00             3
998  e0f9dfd1726c      Там Ем    Tuesday        17:30             1
999  401d5fa8e55d      Там Ем  Wednesday        20:30             1

[1000 rows x 5 columns]

In [243]:
bo_data.duplicated().sum()

0

Скорее всего, такие полные дубликаты говорят о сбое при выгрузке данных. Необходимо сообщить об этом.

Еще одна проблема, о которой предупреждал менеджер проекта — дубликаты карточки Iskra Bar с неправильными названием бара - "Искра" и "Искр". На них заводились реальные заказы, поэтому необходимо вычислить их и переименовать на правильное название — Iskra Bar.

Объявим функцию `replace_restaurant()`, которая заменяет неправильное название, а потом проверяет успешность выполнения замены подсчетом (если замена прошла успешно — результат подсчета должен быть 0).

In [237]:
def replace_restaurant(data, name, replace_name):
    data.replace(name, replace_name, inplace=True)
    return data[data['restaurant']==name]['restaurant'].count()

Передаем функции названия 'Искра' и 'Искр'.

In [241]:
replace_restaurant(bo_data, 'Искра', 'Iskra Bar')

0

In [242]:
replace_restaurant(bo_data, 'Искр', 'Iskra Bar')

0

In [244]:
# дополнительная проверка на дубликаты после замены некорректных названии
bo_data.duplicated().sum()

0

### Выводы

* Название колонок содержат пробелы, что вызвало ошибку при обращении к ним.
* Найдено 65 пропущенных значении типа NaN, Null
* Найдено 24 полных дубликатов
* Некоторые записи содержат некорректные данные, например, название того же ресторана записынный с ошибками, на кириллице, что может привести к появлению дублированных записей отличающими только названием ресторана.
* После замены таких записей и проверки замены дополнительных дубликатов не найдено.

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Тут тоже все отлично)

## Распределение заказов в течение недели

Считаем, на какие кафе приходится больше всего заказов.

In [245]:
bo_data.groupby('restaurant').count().sort_values(by='user_id', ascending=False)

user_id  weekday  ordered_time  seats_number
restaurant                                              
Iskra Bar       287      287           287           287
Там Ем          200      200           200           200
Бар Сибирь      159      159           159           159
Северия         107      107           107           107
Крафтейнер       78       78            78            78
Коворкафе        66       66            66            66
Азия             62       62            62            62
admin            41       41            41            41

Группируем по `weekday` и считаем, на какие дни приходится больше всего заказов.

In [246]:
bo_data.groupby('weekday').count().sort_values(by='user_id', ascending=False)

user_id  restaurant  ordered_time  seats_number
weekday                                                   
Tuesday        251         251           251           251
Wednesday      242         242           242           242
Thursday       180         180           180           180
Monday         141         141           141           141
Friday         110         110           110           110
Saturday        52          52            52            52
Sunday          24          24            24            24

Объявляем функцию, которая возвращает количество заказов для ресторана в определенный день.

In [247]:
def order_count(data, restaurant, weekday):
    count = data[ (data.restaurant.str.lower() == restaurant.lower()) &
                  (data.weekday.str.lower() == weekday.lower()) ].count().user_id
    return count

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Круто!
Вот еще пример функции в две строки:

`def rest_weekday(dataset, rest, weekday):  
    return dataset[(dataset['restaurant'] == rest) & (dataset['weekday'] == weekday)].count()['user_id']`

Передадим ей заведения Iskra Bar и Там Ем — два флагманских места, по которым просили результаты прежде всего — и понедельник, среду и пятницу в качестве дня недели.

In [248]:
order_count(bo_data, 'Iskra Bar', 'monday')

48

In [249]:
order_count(bo_data, 'iskra bar', 'wednesday')

53

In [250]:
order_count(bo_data, 'iskra bar', 'friday')

29

In [251]:
order_count(bo_data, 'там ем', 'monday')

25

In [252]:
order_count(bo_data, 'там ем', 'wednesday')

52

In [253]:
order_count(bo_data, 'там ем', 'friday')

24

Сведем результат в таблицу.

In [1]:
df_orders = bo_data[
        ((bo_data.restaurant.str.lower() == 'iskra bar') | (bo_data.restaurant.str.lower() == 'там ем')) &
        ((bo_data.weekday.str.lower() == 'monday') | (bo_data.weekday.str.lower() == 'wednesday') | (bo_data.weekday.str.lower() == 'friday'))
        ][['restaurant','weekday','seats_number']]
df_orders
df_orders.groupby(['restaurant','weekday']).agg({'seats_number':'count'})


NameError: name 'bo_data' is not defined

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Здесь еще можно сделать pivot таблицу, немного лаконичнее:

`table = bo_data.pivot_table(index = ['restaurant', 'weekday'], values = 'user_id', aggfunc = 'count')
table.query('weekday == ["Monday", "Wednesday", "Friday"] and restaurant == ["Iskra Bar", "Там Ем"]').transpose()`

### Выводы

* По данным больше всего количество заказов приходится на ресторан Iskra Bar - 287
* Больше всего заказов поступает во вторник - 251
* Проанализировав два ресторана с большим колчеством заказов и взяв 3 дня недели понедельник, среда, пятница получили результат:
> * Iskra Bar - больше всего заказов приходится на среду 53, наименьшее количество заказов в пятницу.
> * Там Ем - больше всего заказов приходится на среду 52, наименьшее количество заказов в пятницу.
> * В понедельник Iskra Bar получает почти в два раза больше заказов чем ресторан Там Ем.

## Распределение заказов в течение дня

Сравним количества заказов, сделанных в обеденное и вечернее время — с 12 до 16 и с 16 до 22. Для этого объявим функцию, которая принимает данные и 2 аргумента времени.

In [180]:
def order_count_by_time(data, time_from, time_to):
    import datetime as dt
    
    df_time_range = data[ (pd.to_datetime(data.ordered_time,format='%H:%M').dt.time >= dt.datetime.strptime(time_from,'%H:%M').time()) &
                          (pd.to_datetime(data.ordered_time,format='%H:%M').dt.time <= dt.datetime.strptime(time_to,'%H:%M').time()) ]
    return df_time_range.count().user_id
    

Передадим промежуток с 12:00 и 16:00

In [290]:
order_count_by_time(bo_data, '12:00', '16:00')

588

И с 16:00 до 22:00

In [291]:
order_count_by_time(bo_data, '16:00', '22:00')

338

### Выводы

* Больше всего заказов приходится на обеденное время в промежутке с 12:00 - 16:00 - 588
* В вечернее время в промежутке с 16:00 - 22:00 количество заказов сравнимо меньше - 338

## Типичное количество человек за столиком

Выясним, для какого количества человек обычно заказывают столик в разное время на примере Iskra Bar и Там Ем. Для этого напишем функцию, которая принимает данные, название ресторана, 2 времени и считает моду.

In [169]:
def table_seats_mode(data, restaurant, time_from, time_to):
    df_time_range = data[ (pd.to_datetime(data.ordered_time,format='%H:%M').dt.time >= dt.datetime.strptime(time_from,'%H:%M').time()) &
                          (pd.to_datetime(data.ordered_time,format='%H:%M').dt.time <= dt.datetime.strptime(time_to,'%H:%M').time()) ]
    return df_time_range[df_time_range.restaurant.str.lower() == restaurant.lower()].seats_number.mode() 

Проверим функцию на двух ресторанах, а также дневном и вечернем промежутке времени.

In [170]:
table_seats_mode(bo_data, 'iskra bar', '12:00', '16:00')

0    2
dtype: int64

In [177]:
table_seats_mode(bo_data, 'iskra bar', '16:00', '22:00')

0    1
dtype: int64

In [187]:
table_seats_mode(bo_data, 'там ем', '12:00', '16:00')

0    2
dtype: int64

In [188]:
table_seats_mode(bo_data, 'там ем', '16:00', '22:00')

0    1
dtype: int64

### Выводы

* Были проаналищированы два наипопулярнейших ресторана - Iskra Bar, Там Ем
* Как видно по результатам, обычно заказывают столы на 1, 2 количества человек
* В обеденное время в промежутке с 12:00 - 16:00 чаще всего заказывают стол на 2х
* В вечернее время в промежутке с 16:00 - 22:00 чаще всего заказывают стол на 1 человека.

## Общие выводы

* Три наипопулярных ресторанов на данной территории - Iskra Bar, Там Ем, Бар Сибирь.
* Наибольшее количество заказов 287 имеет ресторан Iskra Bar - 287 заказов. На втором месте ресторан Там Ем - 200. На третьем Бар Сибирь - 159.
* Больше всего заказов в неделе приходится на вторник и среду. Меньше всего заказов приходится на субботу и воскресенье.
* Проанализировав время сделанных заказов делам вывод что чаще всего сервисом пользуются в обеденное время в промежутке с 12:00 - 16:00.
* В рестораны чаще всего приходят небольшими компаниями в обед 2 человека, а в вечернее время чаще всего заказывают стол на 1го (что то даже немного грусно). Можно сделать вывод что чаще всего вечером приходят быстро поужинать в одиночестве потому что не хочется готовить дома :)

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Замечательная работа!